src/main/scala/Hash.scala

```diff
@@ -1,19 +1,643 @@
 package chess
 
-import java.security.MessageDigest
+class Hash(
+    size: Int) {
 
-object Hash {
+  private def roleIndex(role: Role) = role match {
+    case Pawn   => 0
+    case Knight => 1
+    case Bishop => 2
+    case Rook   => 3
+    case Queen  => 4
+    case King   => 5
+  }
+
+  private def pieceIndex(piece: Piece) = piece.color match {
+    case Black => roleIndex(piece.role) * 2
+    case White => roleIndex(piece.role) * 2 + 1
+  }
+
+  private def posIndex(pos: Pos) =
+    8 * (pos.y - 1) + (pos.x - 1)
+
+  private def actorIndex(actor: Actor) =
+    64 * pieceIndex(actor.piece) + posIndex(actor.pos)
+
+  private def crazyPocketIndex(role: Role, color: Color, count: Int) = {
+    assert(0 < count && count <= 16)
+    assert(roleIndex(role) < 5)
+
+    color match {
+      case Black => 16 * roleIndex(role) + (count - 1)
+      case White => 16 * roleIndex(role) + (count - 1) + 16 * 5
+    }
+  }
+
+  private[chess] def hexToBytes(str: String): PositionHash = {
+    str.grouped(2).map(cc =>
+      (Character.digit(cc(0), 16) << 4 | Character.digit(cc(1), 16)).toByte
+    ).toArray take size
+  }
+
+  def apply(situation: Situation): PositionHash = {
+    val board = situation.board
+    val stm = situation.color
+
+    val actors = board.actors.values.map { a =>
+      actorMasks(actorIndex(a))
+    }
+
+    val turn = stm match {
+      case Black => Array(zeroMask)
+      case White => Array(whiteTurnMask)
+    }
+
+    val castling = (situation.history.castles.toList zip castlingMasks).map {
+      case (canCastle, castlingMask) =>
+        if (canCastle) castlingMask else zeroMask
+    }
+
+    // According to FIDE rules en-passant needs to be completely legal in order
+    // to make a difference with regards to threefold repetition.
+    val ep = situation.moves.values.flatten.find(_.enpassant) match {
```

Comment:
I wonder if we could at least reuse the moves generated here, instead of doing-it on a throw-away Situation.